### PR TITLE
Added multi-channel version of felzenszwalb

### LINKED ISSUE #2056: `min_size` is not strictly conformed in the implementation of felzenszwalb
## Description
With `min_size` specified, there're still some segments with sizes that less than it. I don't know if it is an inherent flaw of the algorithm.

## Way to reproduce

    >>> I = skimage.io.imread('dragonbaby.jpg')
    >>> fz = felzenszwalb(I, scale=300, sigma=0.8, min_size=80)
    >>> (fz==9).sum()
    1

dragonbaby

### PR DESCRIPTION
**Pull request by jmetz**
## Description
Adds proper multi-channel support to felzenszwalb, via `_felzenszwalb_rgb`

## References
Fixes #2056

### CODE REVIEW

**Comment by jmetz** [comment]
Just realized I blurred all channels together (i.e. sigma will act across channels) which is not desired; updating to blur channels individually.

File: skimage/segmentation/_felzenszwalb_cy.pyx [hunk 1]
```diff
@@ -125,3 +125,118 @@ def _felzenszwalb_grey(image, double scale=1, sigma=0.8,
...
+
+    Produces an oversegmentation of a multi-channel image using a fast, minimum spanning
+    tree based clustering on the image grid.
+    The number of produced segments as well as their size can only be
+    controlled indirectly through ``scale``. Segment size within an image can
+    vary greatly depending on local contrast.
+
+    Parameters
+    ----------
+    image: ndarray
```
Comment (sciunto) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: add a space between the variable and the column -> image : ndarray. + Elsewhere.

Comment (jmetz) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: hi @sciunto  - I kept the same format for this as the original (`...grey`) implementation. Could you confirm that the 
```
name : description 
```
format is indeed preferred over 
```
name: description
```
please? If so, I guess I could simultaneously fix this in the `...grey` version also.

Comment (soupault) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: @jmetz yes, it is.
See  https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections section `4. Parameters`.

Comment (sciunto) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: I confirm. Thanks for also correcting other mistakes :)

Comment (jmetz) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: Thanks both for the confirmation - I fixed and the original and updated the rgb version also.

File: skimage/segmentation/_felzenszwalb_cy.pyx [hunk 2]
```diff
@@ -125,3 +125,118 @@ def _felzenszwalb_grey(image, double scale=1, sigma=0.8,
...
+        Integer mask indicating segment labels.
+    """
+    if image.ndim != 3:
+        raise ValueError("This algorithm works only on multi-channel 3d"
+                "images. Got image of shape %s" % str(image.shape))
+
+    image = img_as_float(image)
+
+    # rescale scale to behave like in reference implementation
+    scale = float(scale) / 255.
```
Comment (sciunto) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: Are you making the assumption that the image is coded over 8 bits?

File: skimage/segmentation/_felzenszwalb_cy.pyx [hunk 3]
```diff
@@ -125,3 +125,118 @@ def _felzenszwalb_grey(image, double scale=1, sigma=0.8,
         flat = flat[flat]
     flat = np.unique(flat, return_inverse=True)[1]
     return flat.reshape((height, width))
+
+
+def _felzenszwalb_rgb(image, double scale=1, sigma=0.8,
```
Comment (sciunto) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: why not double sigma=0.8?

**Comment by sciunto** [comment]
Actually, it looks it duplicates significantly `_felzenszwalb_grey`. Is there a way to avoid that? like by sharing a common part? It would ease a lot the future maintenance.

File: skimage/segmentation/_felzenszwalb_cy.pyx (hunk 2 above)
Comment (jmetz) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: @sciunto : I kept the same assumptions as in the original algorithm

File: skimage/segmentation/_felzenszwalb_cy.pyx (hunk 3 above)
Comment (jmetz) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: @sciunto : Sigma is only used to pass into `ndimage.gaussian_filter`, consequently there should be no real gain to typing it.

**Comment by jmetz** [comment]
@sciunto : Regarding duplicating large parts of `_felzenszwalb_grey` - this is true, as the 2006 modification made by Felzenszwalb are essentially just in the cost function; in terms of splitting the felzenszwalb function (for example into a function like `generate_cost` and `segment_cost`, this could presumably be done and would likely only add a small overhead to the call time. 

Alternatively, the original function could be entirely removed if we always ensure that the input matrix is  3d - as then `np.abs` is equivalent to `np.sqrt(np.sum( x**2, axis=-1))` though possibly slightly slower. 

Any thoughts as to which would be preferred?

**Comment by sciunto** [comment]
Let's see how other members see this question, but I would say that on the python side, only one function can be presented and it could handle both rgb and grayscale images. Usually, we also have the policy to prefer easy maintenance instead of a small gain of performance.

File: skimage/segmentation/_felzenszwalb_cy.pyx (hunk 2 above)
Comment (sciunto) [inline skimage/segmentation/_felzenszwalb_cy.pyx]: I'm not familiar with this algorithm, but as a user point of view, presented like that, I would feel uncomfortable to pass a 16 bits image for instance :) Either we need to generalize or we should state in the docstring the limits (if it really exists/matters).

**Comment by jni** [comment]
I second @sciunto's opinion.

**Comment by jmetz** [comment]
@sciunto, @jni - Removed gray version of algorithm, instead simply expanding 2d inputs to 3d, as algorithm is agnostic to the actual size of the 3rd dimension.

**Comment by sciunto** [comment]
@jmetz why do you have all these commits? I think you didn't rebase correctly.

**Comment by jmetz** [comment]
@sciunto - I pulled the current master and then checked out my feature branch and did:
```
git rebase master
```
was that not right?

**Comment by sciunto** [comment]
you should not pull master in your branch.

```
git co master
git pull upstream master
git co my_branch
git rebase -i master
```

**Comment by sciunto** [comment]
See https://github.com/scikit-image/scikit-image/blob/master/.github/CONTRIBUTING.txt

**Comment by jmetz** [comment]
Correct - I did not pull master into my feature branch - I pulled master into my master branch and then checked out my feature branch and rebased ... but I did not include the `-i` flag...

**Comment by jmetz** [comment]
I've created a new branch with a clean set of changes and a single commit; add PR.

**Pull request closed without merging**